pkg/api/registered/registered.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package to keep track of API Versions that should be registered in api.Scheme.
+// This is adapted from upstream for use in client/server version negotiation.
+package registered
+
+import (
+	"os"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+// List of registered API versions.
+// The list is in the order of most preferred to the least.
+var RegisteredVersions []string
+
+func init() {
+	validAPIVersions := map[string]bool{
+		"v1": true,
+	}
+
+	// The default list of supported api versions, in order of most preferred to the least.
+	defaultSupportedVersions := "v1"
+	// Env var KUBE_API_VERSIONS is a comma separated list of API versions that should be registered in the scheme.
+	// The versions should be in the order of most preferred to the least.
+	supportedVersions := os.Getenv("OPENSHIFT_API_VERSIONS")
+	if supportedVersions == "" {
+		supportedVersions = defaultSupportedVersions
+	}
+	versions := strings.Split(supportedVersions, ",")
+	for _, version := range versions {
+		// Verify that the version is valid.
+		valid, ok := validAPIVersions[version]
+		if !ok || !valid {
+			// Not a valid API version.
+			glog.Fatalf("invalid api version: %s in OPENSHIFT_API_VERSIONS: %s. List of valid API versions: %v",
+				version, os.Getenv("OPENSHIFT_API_VERSIONS"), validAPIVersions)
+		}
+		RegisteredVersions = append(RegisteredVersions, version)
+	}
+}
+
+// Returns true if the given api version is one of the registered api versions.
+func IsRegisteredAPIVersion(version string) bool {
```

Comment:
Regarding the entirety of `registered.go`: cargo-culted from upstream. I'm in no position to scrutinze the upstream design at this point. If you want, I can just hard-code the API versions at the call site in the factory or whatever. I have no idea if we need the same mechanism upstream's using.